### PR TITLE
Handle tablet proximity

### PR DESCRIPTION
**Pull request by CandyFace**
By using tablet proximity we can assume that while entered, the tablet is in use, which should avoid most weird cases where a mouse event would appear.

This should also fix some cases where the tablet would not register pressure because it didn't get registered as a tablet properly. To clarify what I'm referring to here is the fact that sometimes, when leaving the proximity zone and switching between application, Pencil2D wouldn't register the tablet properly.

Prior to this PR I was able to reproduce a bunch of mouse move events when dabbing *furiously* on my drawing tablet. This is no longer the case.

### CODE REVIEW

**Comment by chchwy** [comment]
sounds like a good idea. will test later on.

**Review by scribblemaniac — CHANGES_REQUESTED** [review]
After testing these changes, I can say that this implementation has issues and that I don't think this is a good approach anyway. It seems that not all tablets will produce tablet proximity events. My Huion H430P on Ubuntu is one such tablet. I suspect that it doesn't even have the capability to produce these events because it does not detect a hovering stylus. This causes major issues with this implementation because mTabletInUse is set to true in the TabletPress event, but is not set to false in the TabletRelease event. And of course re-adding that tablet release event would mean that it could be false while the stylus is still in proximity for those tablets which support that. There are ways around this I'm sure, such as perhaps keeping track of how the tablet usage was detected, but I feel like that is becoming unnecessarily complex.

Ignoring the mouseMoveEvent entirely when the mouse is not active was definitely a mistake. However, I am not yet convinced that this is how we should fix it. Without testing, and without @Jose-Moreno testing, my first instinct is that perhaps we can just skip the mStrokeManager->pointerMoveEvent call if the workaround conditional is true. I would like to hear your thoughts on this @CandyFace.

**Comment by CandyFace** [comment]
Interesting, I definitely thought that all tablets had a proximity sensor, your huion tablet included. I find it odd because my own K58 is a much older model and does produce such events, actually all tablets that I know that has drivers installed, has proximity detection.

Could this be a platform specific issue? 

I did get Jose to test it though and it worked as expected but if we can't assume that all tablets will produce proximity events, then we need to rethink the solution.

File: app/src/pencilapplication.cpp [hunk 1]
```diff
@@ -42,13 +42,18 @@ bool PencilApplication::event(QEvent* event)
...
+    else if (event->type() == QEvent::FileOpen)
     {
         mStartPath = static_cast<QFileOpenEvent*>(event)->file();
         emit openFileRequested(mStartPath);
         return true;
     }
+    else if (event->type() == QEvent::TabletEnterProximity ||
+        event->type() == QEvent::TabletLeaveProximity ) {
+
+        mMainWindow->onTabletProximity(static_cast<QTabletEvent *>(event));
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply app/src/pencilapplication.cpp:54]: Right I admin didn't put much thought into that, I agree with your comment though that being able to handle the event might be a good idea, thus I move the event handling into mainwindow2.

File: core_lib/src/interface/scribblearea.h [hunk 2]
```diff
@@ -170,6 +170,9 @@ public slots:
     void pointerMoveEvent(PointerEvent*);
     void pointerReleaseEvent(PointerEvent*);
 
+    void onTabletLeaveProximity(QTabletEvent *event);
+    void onTabletEnterProximity(QTabletEvent* event);
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/interface/scribblearea.h:174]: That darn OCD :P I will fix.

**Comment by scribblemaniac** [comment]
Okay, I was mistaken, my tablet does have a proximity sensor. I forgot about it because the indicator is the most minuscule light possible and it's only visible if you are looking right over top of the tablet. However that doesn't change the fact that it doesn't produce the events. I would not say it's platform specific because my Wacom tablet produces these events just fine on the same computer. Could be something at either the driver used for Huion tablets, or the way Qt handles it.

**Comment by chchwy** [comment]
I have a feeling we might have to extensively test against major brands of drawing tablets as well as old Qt versions. I saw a bug report about this event wasn't sent in Qt 5.6 and the issue fixed in Qt 5.9.

Ref: https://bugreports.qt.io/browse/QTBUG-53628

Once we decide to relay on this event, we might have to drop the support for legacy systems.

**Comment by CandyFace** [comment]
Hmm right yes it does seem to fragile to be a replacement.
I will undo those changes.

**Pull request closed without merging**